bundle/filefs.go

```diff
@@ -0,0 +1,73 @@
+// +build go1.16
+
+package bundle
+
+import (
+	"fmt"
+	"io"
+	"io/fs"
+	"sync"
+)
+
+const (
+	defaultFSLoaderRoot = "."
+)
+
+type dirLoaderFS struct {
+	sync.Mutex
+	filesystem fs.FS
+	files      []string
+	idx        int
+}
+
+// NewFSLoader returns a basic DirectoryLoader implementation
+// that will load files from a fs.FS interface
+func NewFSLoader(filesystem fs.FS) (DirectoryLoader, error) {
+	d := dirLoaderFS{
+		filesystem: filesystem,
+	}
+
+	err := fs.WalkDir(d.filesystem, defaultFSLoaderRoot, d.walkDir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list files: %w", err)
+	}
+
+	return &d, nil
+}
+
+func (d *dirLoaderFS) walkDir(path string, dirEntry fs.DirEntry, err error) error {
+	if err != nil {
+		return err
+	}
+
+	if dirEntry != nil && dirEntry.Type().IsRegular() {
+		d.files = append(d.files, path)
+	}
+
+	return nil
+}
+
+// NextFile iterates to the next file in the directory tree
+// and returns a file Descriptor for the file.
+func (d *dirLoaderFS) NextFile() (*Descriptor, error) {
+	d.Lock()
+	defer d.Unlock()
+
+	// If done reading files then just return io.EOF
+	// errors for each NextFile() call
+	if d.idx >= len(d.files) {
+		return nil, io.EOF
+	}
+
+	fileName := d.files[d.idx]
+	d.idx++
```

Comment:
This incrementing is the only bit of this method that needs to be synchronized. Thus, it is probably better to use `atomic.AddUint32` since that will avoid the mutex alltogether. In order for this to work, the addition must go at the top of the method, before it is used.